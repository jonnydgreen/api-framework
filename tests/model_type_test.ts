// Copyright 2024-2024 the API framework authors. All rights reserved. MIT license.

import { assert } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import { beforeAll, describe, it } from "@std/testing/bdd";

const stringTestCases: TestCase[] = [
  {
    decoratorType: { imports: [], type: "String" },
    fieldType: { imports: [], type: "number" },
  },
  {
    decoratorType: { imports: [], type: "String" },
    fieldType: { imports: [], type: "boolean" },
  },
  {
    decoratorType: { imports: [], type: "String" },
    fieldType: { imports: [], type: "symbol" },
  },
  {
    decoratorType: { imports: [], type: "String" },
    fieldType: { imports: [], type: "unknown" },
  },
  {
    decoratorType: { imports: [], type: "String" },
    fieldType: { imports: [], type: "undefined" },
  },
  {
    decoratorType: { imports: [], type: "String" },
    fieldType: { imports: [], type: "null" },
  },
];

const numberTestCases: TestCase[] = [
  {
    decoratorType: { imports: [], type: "Number" },
    fieldType: { imports: [], type: "string" },
  },
  {
    decoratorType: { imports: [], type: "Number" },
    fieldType: { imports: [], type: "boolean" },
  },
  {
    decoratorType: { imports: [], type: "Number" },
    fieldType: { imports: [], type: "symbol" },
  },
  {
    decoratorType: { imports: [], type: "Number" },
    fieldType: { imports: [], type: "unknown" },
  },
  {
    decoratorType: { imports: [], type: "Number" },
    fieldType: { imports: [], type: "undefined" },
  },
  {
    decoratorType: { imports: [], type: "Number" },
    fieldType: { imports: [], type: "null" },
  },
];

const booleanTestCases: TestCase[] = [
  {
    decoratorType: { imports: [], type: "Boolean" },
    fieldType: { imports: [], type: "string" },
  },
  {
    decoratorType: { imports: [], type: "Boolean" },
    fieldType: { imports: [], type: "number" },
  },
  {
    decoratorType: { imports: [], type: "Boolean" },
    fieldType: { imports: [], type: "symbol" },
  },
  {
    decoratorType: { imports: [], type: "Boolean" },
    fieldType: { imports: [], type: "unknown" },
  },
  {
    decoratorType: { imports: [], type: "Boolean" },
    fieldType: { imports: [], type: "undefined" },
  },
  {
    decoratorType: { imports: [], type: "Boolean" },
    fieldType: { imports: [], type: "null" },
  },
];

describe("Model types", {
  permissions: {
    read: true,
    run: true,
    write: true,
  },
}, () => {
  describe("String", () => handleTestCases(stringTestCases));
  describe("Number", () => handleTestCases(numberTestCases));
  describe("Boolean", () => handleTestCases(booleanTestCases));
});

function handleTestCases(
  testCases: TestCase[],
): void {
  const snippetResultsMap = new Map<string, Required<TestCase>>();

  beforeAll(async () => {
    await Promise.all(
      testCases.map(async ({ decoratorType, fieldType }) => {
        const snippet = `
                  import { Field, ObjectType } from "@eyrie/app";
                  @ObjectType({ description: "Model" })
                  class Model {
                    @Field({ description: "Field", type: ${decoratorType.type} })
                    field!: ${fieldType.type}
                  }`;
        snippetResultsMap.set(`${decoratorType.type}-${fieldType.type}`, {
          decoratorType,
          fieldType,
          results: await evaluateSnippet(snippet),
        });
      }),
    );
  });

  testCases.forEach(({ decoratorType, fieldType }) => {
    it(`should error model field decorator of type '${decoratorType.type}' does not map to field type of type '${fieldType.type}'`, async (t) => {
      // Act
      const key = `${decoratorType.type}-${fieldType.type}`;
      const snippetResults = snippetResultsMap.get(key);
      assert(snippetResults, `No results found key snippet key: ${key}`);
      const diagnostics = sanitiseDiagnostics(
        snippetResults.results.diagnostics,
      );

      // Assert
      assert(
        !snippetResults.results.success,
        `Snippet did not fail to compile: ${snippetResults.results.snippet}`,
      );
      await assertSnapshot(t, diagnostics);
    });
  });
}

interface SnippetResult {
  snippet: string;
  success: boolean;
  diagnostics: string;
}

interface TestCase {
  decoratorType: { imports: string[]; type: string };
  fieldType: { imports: string[]; type: string };
  results?: SnippetResult;
}

function sanitiseDiagnostics(diagnostics: string): string {
  return diagnostics.replace(
    /file:\/\/.+\/\$deno\$eval\.ts/g,
    "file://__PATH__/$deno$eval.ts",
  );
}

async function evaluateSnippet(
  snippet: string,
): Promise<SnippetResult> {
  const command = new Deno.Command(Deno.execPath(), {
    args: [
      "eval",
      "--ext=ts",
      "--check",
      "--unstable-webgpu",
      "--no-lock",
      snippet,
    ],
    env: {
      NO_COLOR: "true",
    },
    stderr: "piped",
  });
  const { success, stderr } = await command.output();
  const diagnostics = new TextDecoder().decode(stderr);
  return { snippet: snippet, success, diagnostics };
}
