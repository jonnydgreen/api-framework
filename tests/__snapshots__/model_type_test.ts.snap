export const snapshot = {};

snapshot[`Model types > String > should error model field decorator of type 'String' does not map to field type of type 'number' 1`] = `
\`Check file://__PATH__/\$deno\$eval.ts
error: TS1240 [ERROR]: Unable to resolve signature of property decorator when called as an expression.
  Argument of type 'ClassFieldDecoratorContext<Model, number> & { name: "field"; private: false; static: false; }' is not assignable to parameter of type 'ClassFieldDecoratorContext<Model | undefined, string>'.
    The types returned by 'access.get(...)' are incompatible between these types.
      Type 'number' is not assignable to type 'string'.
                    @Field({ description: "Field", type: String })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

TS1270 [ERROR]: Decorator function return type '(this: Model | undefined, value: string) => string' is not assignable to type 'void | ((this: Model, value: number) => number)'.
  Type '(this: Model | undefined, value: string) => string' is not assignable to type '(this: Model, value: number) => number'.
    Types of parameters 'value' and 'value' are incompatible.
      Type 'number' is not assignable to type 'string'.
                    @Field({ description: "Field", type: String })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

Found 2 errors.
\`
`;

snapshot[`Model types > String > should error model field decorator of type 'String' does not map to field type of type 'boolean' 1`] = `
\`Check file://__PATH__/\$deno\$eval.ts
error: TS1240 [ERROR]: Unable to resolve signature of property decorator when called as an expression.
  Argument of type 'ClassFieldDecoratorContext<Model, boolean> & { name: "field"; private: false; static: false; }' is not assignable to parameter of type 'ClassFieldDecoratorContext<Model | undefined, string>'.
    The types returned by 'access.get(...)' are incompatible between these types.
      Type 'boolean' is not assignable to type 'string'.
                    @Field({ description: "Field", type: String })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

TS1270 [ERROR]: Decorator function return type '(this: Model | undefined, value: string) => string' is not assignable to type 'void | ((this: Model, value: boolean) => boolean)'.
  Type '(this: Model | undefined, value: string) => string' is not assignable to type '(this: Model, value: boolean) => boolean'.
    Types of parameters 'value' and 'value' are incompatible.
      Type 'boolean' is not assignable to type 'string'.
                    @Field({ description: "Field", type: String })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

Found 2 errors.
\`
`;

snapshot[`Model types > String > should error model field decorator of type 'String' does not map to field type of type 'symbol' 1`] = `
\`Check file://__PATH__/\$deno\$eval.ts
error: TS1240 [ERROR]: Unable to resolve signature of property decorator when called as an expression.
  Argument of type 'ClassFieldDecoratorContext<Model, symbol> & { name: "field"; private: false; static: false; }' is not assignable to parameter of type 'ClassFieldDecoratorContext<Model | undefined, string>'.
    The types returned by 'access.get(...)' are incompatible between these types.
      Type 'symbol' is not assignable to type 'string'.
                    @Field({ description: "Field", type: String })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

TS1270 [ERROR]: Decorator function return type '(this: Model | undefined, value: string) => string' is not assignable to type 'void | ((this: Model, value: symbol) => symbol)'.
  Type '(this: Model | undefined, value: string) => string' is not assignable to type '(this: Model, value: symbol) => symbol'.
    Types of parameters 'value' and 'value' are incompatible.
      Type 'symbol' is not assignable to type 'string'.
                    @Field({ description: "Field", type: String })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

Found 2 errors.
\`
`;

snapshot[`Model types > String > should error model field decorator of type 'String' does not map to field type of type 'unknown' 1`] = `
\`Check file://__PATH__/\$deno\$eval.ts
error: TS1240 [ERROR]: Unable to resolve signature of property decorator when called as an expression.
  Argument of type 'ClassFieldDecoratorContext<Model, unknown> & { name: "field"; private: false; static: false; }' is not assignable to parameter of type 'ClassFieldDecoratorContext<Model | undefined, string>'.
    The types returned by 'access.get(...)' are incompatible between these types.
      Type 'unknown' is not assignable to type 'string'.
                    @Field({ description: "Field", type: String })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

TS1270 [ERROR]: Decorator function return type '(this: Model | undefined, value: string) => string' is not assignable to type 'void | ((this: Model, value: unknown) => unknown)'.
  Type '(this: Model | undefined, value: string) => string' is not assignable to type '(this: Model, value: unknown) => unknown'.
    Types of parameters 'value' and 'value' are incompatible.
      Type 'unknown' is not assignable to type 'string'.
                    @Field({ description: "Field", type: String })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

Found 2 errors.
\`
`;

snapshot[`Model types > String > should error model field decorator of type 'String' does not map to field type of type 'undefined' 1`] = `
\`Check file://__PATH__/\$deno\$eval.ts
error: TS1240 [ERROR]: Unable to resolve signature of property decorator when called as an expression.
  Argument of type 'ClassFieldDecoratorContext<Model, undefined> & { name: "field"; private: false; static: false; }' is not assignable to parameter of type 'ClassFieldDecoratorContext<Model | undefined, string>'.
    The types returned by 'access.get(...)' are incompatible between these types.
      Type 'undefined' is not assignable to type 'string'.
                    @Field({ description: "Field", type: String })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

TS1270 [ERROR]: Decorator function return type '(this: Model | undefined, value: string) => string' is not assignable to type 'void | ((this: Model, value: undefined) => undefined)'.
  Type '(this: Model | undefined, value: string) => string' is not assignable to type '(this: Model, value: undefined) => undefined'.
    Types of parameters 'value' and 'value' are incompatible.
      Type 'undefined' is not assignable to type 'string'.
                    @Field({ description: "Field", type: String })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

Found 2 errors.
\`
`;

snapshot[`Model types > String > should error model field decorator of type 'String' does not map to field type of type 'null' 1`] = `
\`Check file://__PATH__/\$deno\$eval.ts
error: TS1240 [ERROR]: Unable to resolve signature of property decorator when called as an expression.
  Argument of type 'ClassFieldDecoratorContext<Model, null> & { name: "field"; private: false; static: false; }' is not assignable to parameter of type 'ClassFieldDecoratorContext<Model | undefined, string>'.
    The types returned by 'access.get(...)' are incompatible between these types.
      Type 'null' is not assignable to type 'string'.
                    @Field({ description: "Field", type: String })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

TS1270 [ERROR]: Decorator function return type '(this: Model | undefined, value: string) => string' is not assignable to type 'void | ((this: Model, value: null) => null)'.
  Type '(this: Model | undefined, value: string) => string' is not assignable to type '(this: Model, value: null) => null'.
    Types of parameters 'value' and 'value' are incompatible.
      Type 'null' is not assignable to type 'string'.
                    @Field({ description: "Field", type: String })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

Found 2 errors.
\`
`;

snapshot[`Model types > Number > should error model field decorator of type 'Number' does not map to field type of type 'string' 1`] = `
\`Check file://__PATH__/\$deno\$eval.ts
error: TS1240 [ERROR]: Unable to resolve signature of property decorator when called as an expression.
  Argument of type 'ClassFieldDecoratorContext<Model, string> & { name: "field"; private: false; static: false; }' is not assignable to parameter of type 'ClassFieldDecoratorContext<Model | undefined, number>'.
    The types returned by 'access.get(...)' are incompatible between these types.
      Type 'string' is not assignable to type 'number'.
                    @Field({ description: "Field", type: Number })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

TS1270 [ERROR]: Decorator function return type '(this: Model | undefined, value: number) => number' is not assignable to type 'void | ((this: Model, value: string) => string)'.
  Type '(this: Model | undefined, value: number) => number' is not assignable to type '(this: Model, value: string) => string'.
    Types of parameters 'value' and 'value' are incompatible.
      Type 'string' is not assignable to type 'number'.
                    @Field({ description: "Field", type: Number })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

Found 2 errors.
\`
`;

snapshot[`Model types > Number > should error model field decorator of type 'Number' does not map to field type of type 'boolean' 1`] = `
\`Check file://__PATH__/\$deno\$eval.ts
error: TS1240 [ERROR]: Unable to resolve signature of property decorator when called as an expression.
  Argument of type 'ClassFieldDecoratorContext<Model, boolean> & { name: "field"; private: false; static: false; }' is not assignable to parameter of type 'ClassFieldDecoratorContext<Model | undefined, number>'.
    The types returned by 'access.get(...)' are incompatible between these types.
      Type 'boolean' is not assignable to type 'number'.
                    @Field({ description: "Field", type: Number })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

TS1270 [ERROR]: Decorator function return type '(this: Model | undefined, value: number) => number' is not assignable to type 'void | ((this: Model, value: boolean) => boolean)'.
  Type '(this: Model | undefined, value: number) => number' is not assignable to type '(this: Model, value: boolean) => boolean'.
    Types of parameters 'value' and 'value' are incompatible.
      Type 'boolean' is not assignable to type 'number'.
                    @Field({ description: "Field", type: Number })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

Found 2 errors.
\`
`;

snapshot[`Model types > Number > should error model field decorator of type 'Number' does not map to field type of type 'symbol' 1`] = `
\`Check file://__PATH__/\$deno\$eval.ts
error: TS1240 [ERROR]: Unable to resolve signature of property decorator when called as an expression.
  Argument of type 'ClassFieldDecoratorContext<Model, symbol> & { name: "field"; private: false; static: false; }' is not assignable to parameter of type 'ClassFieldDecoratorContext<Model | undefined, number>'.
    The types returned by 'access.get(...)' are incompatible between these types.
      Type 'symbol' is not assignable to type 'number'.
                    @Field({ description: "Field", type: Number })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

TS1270 [ERROR]: Decorator function return type '(this: Model | undefined, value: number) => number' is not assignable to type 'void | ((this: Model, value: symbol) => symbol)'.
  Type '(this: Model | undefined, value: number) => number' is not assignable to type '(this: Model, value: symbol) => symbol'.
    Types of parameters 'value' and 'value' are incompatible.
      Type 'symbol' is not assignable to type 'number'.
                    @Field({ description: "Field", type: Number })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

Found 2 errors.
\`
`;

snapshot[`Model types > Number > should error model field decorator of type 'Number' does not map to field type of type 'unknown' 1`] = `
\`Check file://__PATH__/\$deno\$eval.ts
error: TS1240 [ERROR]: Unable to resolve signature of property decorator when called as an expression.
  Argument of type 'ClassFieldDecoratorContext<Model, unknown> & { name: "field"; private: false; static: false; }' is not assignable to parameter of type 'ClassFieldDecoratorContext<Model | undefined, number>'.
    The types returned by 'access.get(...)' are incompatible between these types.
      Type 'unknown' is not assignable to type 'number'.
                    @Field({ description: "Field", type: Number })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

TS1270 [ERROR]: Decorator function return type '(this: Model | undefined, value: number) => number' is not assignable to type 'void | ((this: Model, value: unknown) => unknown)'.
  Type '(this: Model | undefined, value: number) => number' is not assignable to type '(this: Model, value: unknown) => unknown'.
    Types of parameters 'value' and 'value' are incompatible.
      Type 'unknown' is not assignable to type 'number'.
                    @Field({ description: "Field", type: Number })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

Found 2 errors.
\`
`;

snapshot[`Model types > Number > should error model field decorator of type 'Number' does not map to field type of type 'undefined' 1`] = `
\`Check file://__PATH__/\$deno\$eval.ts
error: TS1240 [ERROR]: Unable to resolve signature of property decorator when called as an expression.
  Argument of type 'ClassFieldDecoratorContext<Model, undefined> & { name: "field"; private: false; static: false; }' is not assignable to parameter of type 'ClassFieldDecoratorContext<Model | undefined, number>'.
    The types returned by 'access.get(...)' are incompatible between these types.
      Type 'undefined' is not assignable to type 'number'.
                    @Field({ description: "Field", type: Number })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

TS1270 [ERROR]: Decorator function return type '(this: Model | undefined, value: number) => number' is not assignable to type 'void | ((this: Model, value: undefined) => undefined)'.
  Type '(this: Model | undefined, value: number) => number' is not assignable to type '(this: Model, value: undefined) => undefined'.
    Types of parameters 'value' and 'value' are incompatible.
      Type 'undefined' is not assignable to type 'number'.
                    @Field({ description: "Field", type: Number })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

Found 2 errors.
\`
`;

snapshot[`Model types > Number > should error model field decorator of type 'Number' does not map to field type of type 'null' 1`] = `
\`Check file://__PATH__/\$deno\$eval.ts
error: TS1240 [ERROR]: Unable to resolve signature of property decorator when called as an expression.
  Argument of type 'ClassFieldDecoratorContext<Model, null> & { name: "field"; private: false; static: false; }' is not assignable to parameter of type 'ClassFieldDecoratorContext<Model | undefined, number>'.
    The types returned by 'access.get(...)' are incompatible between these types.
      Type 'null' is not assignable to type 'number'.
                    @Field({ description: "Field", type: Number })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

TS1270 [ERROR]: Decorator function return type '(this: Model | undefined, value: number) => number' is not assignable to type 'void | ((this: Model, value: null) => null)'.
  Type '(this: Model | undefined, value: number) => number' is not assignable to type '(this: Model, value: null) => null'.
    Types of parameters 'value' and 'value' are incompatible.
      Type 'null' is not assignable to type 'number'.
                    @Field({ description: "Field", type: Number })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

Found 2 errors.
\`
`;

snapshot[`Model types > Boolean > should error model field decorator of type 'Boolean' does not map to field type of type 'string' 1`] = `
\`Check file://__PATH__/\$deno\$eval.ts
error: TS1240 [ERROR]: Unable to resolve signature of property decorator when called as an expression.
  Argument of type 'ClassFieldDecoratorContext<Model, string> & { name: "field"; private: false; static: false; }' is not assignable to parameter of type 'ClassFieldDecoratorContext<Model | undefined, boolean>'.
    The types returned by 'access.get(...)' are incompatible between these types.
      Type 'string' is not assignable to type 'boolean'.
                    @Field({ description: "Field", type: Boolean })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

TS1270 [ERROR]: Decorator function return type '(this: Model | undefined, value: boolean) => boolean' is not assignable to type 'void | ((this: Model, value: string) => string)'.
  Type '(this: Model | undefined, value: boolean) => boolean' is not assignable to type '(this: Model, value: string) => string'.
    Types of parameters 'value' and 'value' are incompatible.
      Type 'string' is not assignable to type 'boolean'.
                    @Field({ description: "Field", type: Boolean })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

Found 2 errors.
\`
`;

snapshot[`Model types > Boolean > should error model field decorator of type 'Boolean' does not map to field type of type 'number' 1`] = `
\`Check file://__PATH__/\$deno\$eval.ts
error: TS1240 [ERROR]: Unable to resolve signature of property decorator when called as an expression.
  Argument of type 'ClassFieldDecoratorContext<Model, number> & { name: "field"; private: false; static: false; }' is not assignable to parameter of type 'ClassFieldDecoratorContext<Model | undefined, boolean>'.
    The types returned by 'access.get(...)' are incompatible between these types.
      Type 'number' is not assignable to type 'boolean'.
                    @Field({ description: "Field", type: Boolean })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

TS1270 [ERROR]: Decorator function return type '(this: Model | undefined, value: boolean) => boolean' is not assignable to type 'void | ((this: Model, value: number) => number)'.
  Type '(this: Model | undefined, value: boolean) => boolean' is not assignable to type '(this: Model, value: number) => number'.
    Types of parameters 'value' and 'value' are incompatible.
      Type 'number' is not assignable to type 'boolean'.
                    @Field({ description: "Field", type: Boolean })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

Found 2 errors.
\`
`;

snapshot[`Model types > Boolean > should error model field decorator of type 'Boolean' does not map to field type of type 'symbol' 1`] = `
\`Check file://__PATH__/\$deno\$eval.ts
error: TS1240 [ERROR]: Unable to resolve signature of property decorator when called as an expression.
  Argument of type 'ClassFieldDecoratorContext<Model, symbol> & { name: "field"; private: false; static: false; }' is not assignable to parameter of type 'ClassFieldDecoratorContext<Model | undefined, boolean>'.
    The types returned by 'access.get(...)' are incompatible between these types.
      Type 'symbol' is not assignable to type 'boolean'.
                    @Field({ description: "Field", type: Boolean })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

TS1270 [ERROR]: Decorator function return type '(this: Model | undefined, value: boolean) => boolean' is not assignable to type 'void | ((this: Model, value: symbol) => symbol)'.
  Type '(this: Model | undefined, value: boolean) => boolean' is not assignable to type '(this: Model, value: symbol) => symbol'.
    Types of parameters 'value' and 'value' are incompatible.
      Type 'symbol' is not assignable to type 'boolean'.
                    @Field({ description: "Field", type: Boolean })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

Found 2 errors.
\`
`;

snapshot[`Model types > Boolean > should error model field decorator of type 'Boolean' does not map to field type of type 'unknown' 1`] = `
\`Check file://__PATH__/\$deno\$eval.ts
error: TS1240 [ERROR]: Unable to resolve signature of property decorator when called as an expression.
  Argument of type 'ClassFieldDecoratorContext<Model, unknown> & { name: "field"; private: false; static: false; }' is not assignable to parameter of type 'ClassFieldDecoratorContext<Model | undefined, boolean>'.
    The types returned by 'access.get(...)' are incompatible between these types.
      Type 'unknown' is not assignable to type 'boolean'.
                    @Field({ description: "Field", type: Boolean })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

TS1270 [ERROR]: Decorator function return type '(this: Model | undefined, value: boolean) => boolean' is not assignable to type 'void | ((this: Model, value: unknown) => unknown)'.
  Type '(this: Model | undefined, value: boolean) => boolean' is not assignable to type '(this: Model, value: unknown) => unknown'.
    Types of parameters 'value' and 'value' are incompatible.
      Type 'unknown' is not assignable to type 'boolean'.
                    @Field({ description: "Field", type: Boolean })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

Found 2 errors.
\`
`;

snapshot[`Model types > Boolean > should error model field decorator of type 'Boolean' does not map to field type of type 'undefined' 1`] = `
\`Check file://__PATH__/\$deno\$eval.ts
error: TS1240 [ERROR]: Unable to resolve signature of property decorator when called as an expression.
  Argument of type 'ClassFieldDecoratorContext<Model, undefined> & { name: "field"; private: false; static: false; }' is not assignable to parameter of type 'ClassFieldDecoratorContext<Model | undefined, boolean>'.
    The types returned by 'access.get(...)' are incompatible between these types.
      Type 'undefined' is not assignable to type 'boolean'.
                    @Field({ description: "Field", type: Boolean })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

TS1270 [ERROR]: Decorator function return type '(this: Model | undefined, value: boolean) => boolean' is not assignable to type 'void | ((this: Model, value: undefined) => undefined)'.
  Type '(this: Model | undefined, value: boolean) => boolean' is not assignable to type '(this: Model, value: undefined) => undefined'.
    Types of parameters 'value' and 'value' are incompatible.
      Type 'undefined' is not assignable to type 'boolean'.
                    @Field({ description: "Field", type: Boolean })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

Found 2 errors.
\`
`;

snapshot[`Model types > Boolean > should error model field decorator of type 'Boolean' does not map to field type of type 'null' 1`] = `
\`Check file://__PATH__/\$deno\$eval.ts
error: TS1240 [ERROR]: Unable to resolve signature of property decorator when called as an expression.
  Argument of type 'ClassFieldDecoratorContext<Model, null> & { name: "field"; private: false; static: false; }' is not assignable to parameter of type 'ClassFieldDecoratorContext<Model | undefined, boolean>'.
    The types returned by 'access.get(...)' are incompatible between these types.
      Type 'null' is not assignable to type 'boolean'.
                    @Field({ description: "Field", type: Boolean })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

TS1270 [ERROR]: Decorator function return type '(this: Model | undefined, value: boolean) => boolean' is not assignable to type 'void | ((this: Model, value: null) => null)'.
  Type '(this: Model | undefined, value: boolean) => boolean' is not assignable to type '(this: Model, value: null) => null'.
    Types of parameters 'value' and 'value' are incompatible.
      Type 'null' is not assignable to type 'boolean'.
                    @Field({ description: "Field", type: Boolean })
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file://__PATH__/\$deno\$eval.ts:5:22

Found 2 errors.
\`
`;
